File: src/prod/src/managed/Microsoft.ServiceFabric.Data.Impl/Replicator/OperationLogRecord.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace Microsoft.ServiceFabric.Replicator
{
    using System;
    using System.Collections.Generic;
    using System.Fabric;
    using System.Fabric.Common;
    using System.IO;

    internal sealed class OperationLogRecord : TransactionLogRecord
    {
        private static readonly uint ApproximateDiskSpaceUsed = sizeof(int) + (3*sizeof(int)) + sizeof(bool);

        private static readonly int NullOperationDataCode = -1;

        // The below fields are persisted
        private bool isRedoOnly;

        private List<ArraySegment<byte>> replicatedData;

        private OperationData metaData;

        private object operationContext;

        private OperationData redo;

        private OperationData undo;

        internal OperationLogRecord(LogRecordType recordType, ulong recordPosition, long lsn)
            : base(recordType, recordPosition, lsn)
        {
            Utility.Assert(recordType == LogRecordType.Operation, "recordType == LogRecordType.Operation");
            this.metaData = null;
            this.redo = null;
            this.undo = null;
            this.operationContext = null;
            this.isRedoOnly = false;
        }

        internal OperationLogRecord(
            TransactionBase transaction,
            OperationData metaData,
            OperationData undo,
            OperationData redo,
            object operationContext)
            : base(LogRecordType.Operation, transaction, InvalidTransactionLogRecord)
        {
            this.metaData = metaData;
            this.undo = undo;
            this.redo = redo;
            this.operationContext = operationContext;
            this.isRedoOnly = false;

            this.UpdateApproximateDiskSize();
        }

        internal OperationLogRecord(
            TransactionBase transaction,
            OperationData metaData,
            OperationData redo,
            object operationContext)
            : base(LogRecordType.Operation, transaction, InvalidTransactionLogRecord)
        {
            this.metaData = metaData;
            this.undo = null;
            this.redo = redo;
            this.operationContext = operationContext;
            this.isRedoOnly = true;

            this.UpdateApproximateDiskSize();
        }

        internal override bool IsLatencySensitiveRecord
        {
            get { return this.Transaction.IsAtomicOperation; }
        }

        internal bool IsOperationContextPresent
        {
            get { return this.operationContext != null; }
        }

        internal bool IsRedoOnly
        {
            get { return this.isRedoOnly; }
        }

        internal OperationData MetaData
        {
            get { return this.metaData; }

            set { this.metaData = value; }
        }

        internal object OperationContext
        {
            set
            {
                // todo preethas:  With name being added to operation context, needs value comparison and not reference. Removing it for now.
                // Utility.Assert((this.operationContext == null) || (this.operationContext == value));
                this.operationContext = value;
            }
        }

        internal OperationData Redo
        {
            get { return this.redo; }
        }

        internal OperationData Undo
        {
            get { return this.undo; }
        }

        internal static OperationData ReadOperationData(OperationData operationData, ref int index)
        {
            int count;
            using (var br = new BinaryReader(IncrementIndexAndGetMemoryStreamAt(operationData, ref index)))
            {
                count = br.ReadInt32();
            }

            if (count == NullOperationDataCode)
            {
                return null;
            }

            var operationData2 = new OperationData();
            for (var i = 0; i < count; i++)
            {
                operationData2.Add(ReadBytes(operationData, ref index));
            }

            return operationData2;
        }

        internal static OperationData ReadOperationData(BinaryReader br)
        {
            var count = br.ReadInt32();
            if (count == NullOperationDataCode)
            {
                return null;
            }

            var operationData = new OperationData();
            for (var i = 0; i < count; i++)
            {
                operationData.Add(ReadBytes(br));
            }

            return operationData;
        }

        internal static void WriteOperationData(
            OperationData bytes,
            BinaryWriter bw,
            ref List<ArraySegment<byte>> writtenSegments)
        {
            var startingPos = bw.BaseStream.Position;

            if (bytes == null)
            {
                bw.Write(NullOperationDataCode);
                writtenSegments.Add(CreateArraySegment(startingPos, bw));
                return;
            }

            bw.Write(bytes.Count);
            writtenSegments.Add(CreateArraySegment(startingPos, bw));

            foreach (var arraySegment in bytes)
            {
                startingPos = bw.BaseStream.Position;
                bw.Write(arraySegment.Count);
                writtenSegments.Add(CreateArraySegment(startingPos, bw));

                // NOTE: Include the array segment only if it has some data. Otherwise, V1 replicator fails replication
                if (arraySegment.Count > 0)
                {
                    writtenSegments.Add(arraySegment);
                }
            }
        }

        internal object ResetOperationContext()
        {
            var context = this.operationContext;
            if (context != null)
            {
                // context = Interlocked.CompareExchange(ref this.operationContext, null, context);
                this.operationContext = null;
            }

            return context;
        }

        protected override int GetSizeOnWire()
        {
            var size = base.GetSizeOnWire() + sizeof(int) + this.CalculateWireSize(this.metaData) + this.CalculateWireSize(this.redo)
                       + sizeof(bool);

            if (this.isRedoOnly == false)
            {
                size += this.CalculateWireSize(this.undo);
            }

            return size;
        }

        internal override void InvalidateReplicatedDataBuffers(SynchronizedBufferPool<BinaryWriter> pool)
        {
            base.InvalidateReplicatedDataBuffers(pool);
            this.replicatedData = null;
        }

        protected override void Read(BinaryReader br, bool isPhysicalRead)
        {
            // IF THIS METHOD CHANGES, READLOGICAL MUST CHANGE ACCORDINGLY AS WE ARE ADDING MULTIPLE ARRAYSEGMENTS HERE
            // THIS IS TRUE FOR ANY REDO+UNDO TYPE RECORDS
            base.Read(br, isPhysicalRead);

            // Metadata section.
            var startingPosition = br.BaseStream.Position;
            var sizeOfSection = br.ReadInt32();
            var endPosition = startingPosition + sizeOfSection;

            // Read Metadata fields
            this.isRedoOnly = br.ReadBoolean();

            // Jump to the end of the section ignoring fields that are not understood.
            Utility.Assert(endPosition >= br.BaseStream.Position, "Could not have read more than section size.");
            br.BaseStream.Position = endPosition;

            if (this.isRedoOnly)
            {
                this.Transaction = AtomicRedoOperation.CreateAtomicRedoOperation(this.Transaction.Id, true);
            }

            this.metaData = ReadOperationData(br);
            this.redo = ReadOperationData(br);

            if (!this.isRedoOnly)
            {
                this.undo = ReadOperationData(br);
            }

            this.UpdateApproximateDiskSize();
            return;
        }

        protected override void ReadLogical(OperationData operationData, ref int index)
        {
            base.ReadLogical(operationData, ref index);

            using (var br = new BinaryReader(IncrementIndexAndGetMemoryStreamAt(operationData, ref index)))
            {
                // Metadata section.
                var startingPosition = br.BaseStream.Position;
                var sizeOfSection = br.ReadInt32();
                var endPosition = startingPosition + sizeOfSection;

                // Read Metadata fields
                this.isRedoOnly = br.ReadBoolean();

                // Jump to the end of the section ignoring fields that are not understood.
                Utility.Assert(endPosition >= br.BaseStream.Position, "Could not have read more than section size.");
                br.BaseStream.Position = endPosition;
            }

            if (this.isRedoOnly)
            {
                this.Transaction = AtomicRedoOperation.CreateAtomicRedoOperation(this.Transaction.Id, true);
            }

            this.metaData = ReadOperationData(operationData, ref index);
            this.redo = ReadOperationData(operationData, ref index);

            if (!this.isRedoOnly)
            {
                this.undo = ReadOperationData(operationData, ref index);
            }

            this.UpdateApproximateDiskSize();
        }

        protected override void Write(BinaryWriter bw, OperationData operationData, bool isPhysicalWrite, bool forceRecomputeOffsets)
        {
            base.Write(bw, operationData, isPhysicalWrite, forceRecomputeOffsets);

            if (this.replicatedData == null)
            {
                this.replicatedData = new List<ArraySegment<byte>>();
                // Metadata section.
                var startingPos = bw.BaseStream.Position;
                bw.BaseStream.Position += sizeof(int);

                // Write Metadata fields
                bw.Write(this.isRedoOnly);

                // End Metadata section
                var endPosition = bw.BaseStream.Position;
                var sizeOfSection = checked((int) (endPosition - startingPos));
                bw.BaseStream.Position = startingPos;
                bw.Write(sizeOfSection);
                bw.BaseStream.Position = endPosition;

                this.replicatedData.Add(CreateArraySegment(startingPos, bw));
                WriteOperationData(this.metaData, bw, ref this.replicatedData);
                WriteOperationData(this.redo, bw, ref this.replicatedData);

                if (!this.isRedoOnly)
                {
                    WriteOperationData(this.undo, bw, ref this.replicatedData);
                }
            }

            foreach (var segment in this.replicatedData)
            {
                operationData.Add(segment);
            }

            return;
        }

        private static ArraySegment<byte> ReadBytes(OperationData operationData, ref int index)
        {
            int count;
            using (var br = new BinaryReader(IncrementIndexAndGetMemoryStreamAt(operationData, ref index)))
            {
                count = br.ReadInt32();
            }

            if (count == 0)
            {
                return new ArraySegment<byte>(new byte[0]);
            }

            return IncrementIndexAndGetArraySegmentAt(operationData, ref index);
        }

        private static ArraySegment<byte> ReadBytes(BinaryReader br)
        {
            var count = br.ReadInt32();
            var array = br.ReadBytes(count);

            return new ArraySegment<byte>(array);
        }

        private void UpdateApproximateDiskSize()
        {
            if (this.metaData != null)
            {
                foreach (var buffer in this.metaData)
                {
                    this.ApproximateSizeOnDisk += (uint) buffer.Count + sizeof(int);
                }
            }

            if (this.undo != null)
            {
                foreach (var buffer in this.undo)
                {
                    this.ApproximateSizeOnDisk += (uint) buffer.Count + sizeof(int);
                }
            }

            if (this.redo != null)
            {
                foreach (var buffer in this.redo)
                {
                    this.ApproximateSizeOnDisk += (uint) buffer.Count + sizeof(int);
                }
            }

            this.ApproximateSizeOnDisk += ApproximateDiskSpaceUsed;
        }
    }
}

File: src/prod/src/data/txnreplicator/logrecordlib/OperationLogRecord.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

using namespace ktl;
using namespace Data::LogRecordLib;
using namespace TxnReplicator;
using namespace Data::Utilities;

const ULONG OperationLogRecord::DiskSpaceUsed = sizeof(LONG32) + sizeof(bool);

OperationLogRecord::OperationLogRecord(
    __in LogRecordType::Enum recordType,
    __in ULONG64 recordPosition,
    __in LONG64 lsn,
    __in PhysicalLogRecord & invalidPhysicalLogRecord,
    __in TransactionLogRecord & invalidTransactionLog)
    : TransactionLogRecord(recordType, recordPosition, lsn, invalidPhysicalLogRecord, invalidTransactionLog)
    , isRedoOnly_(false)
    , metaData_(nullptr)
    , undo_(nullptr)
    , redo_(nullptr)
    , operationContext_(nullptr)
    , replicatedData_(nullptr)
{
    ASSERT_IFNOT(recordType == LogRecordType::Enum::Operation, "Expected operation log record type, actual: {0}", recordType);
}

OperationLogRecord::OperationLogRecord(
    __in TransactionBase & transaction,
    __in_opt OperationData const * const metaData,
    __in_opt OperationData const * const undo,
    __in_opt OperationData const * const redo,
    __in_opt OperationContext const * const operationContext,
    __in PhysicalLogRecord & invalidPhysicalLogRecord,
    __in TransactionLogRecord & invalidTransactionLog)
    : TransactionLogRecord(LogRecordType::Enum::Operation, transaction, invalidPhysicalLogRecord, invalidTransactionLog, &invalidTransactionLog)
    , isRedoOnly_(false)
    , metaData_(metaData)
    , undo_(undo)
    , redo_(redo)
    , operationContext_(operationContext)
    , replicatedData_(nullptr)
{
    UpdateApproximateDiskSize();
}

OperationLogRecord::OperationLogRecord(
    __in TransactionBase & transaction,
    __in_opt OperationData const * const metaData,
    __in_opt OperationData const * const redo,
    __in_opt OperationContext const * const operationContext,
    __in PhysicalLogRecord & invalidPhysicalLogRecord,
    __in TransactionLogRecord & invalidTransactionLog)
    : TransactionLogRecord(LogRecordType::Enum::Operation, transaction, invalidPhysicalLogRecord, invalidTransactionLog, &invalidTransactionLog)
    , isRedoOnly_(true)
    , metaData_(metaData)
    , undo_(nullptr)
    , redo_(redo)
    , operationContext_(operationContext)
    , replicatedData_(nullptr)
{
    UpdateApproximateDiskSize();
}

OperationLogRecord::~OperationLogRecord()
{
}

OperationLogRecord::SPtr OperationLogRecord::Create(
    __in LogRecordType::Enum recordType,
    __in ULONG64 recordPosition,
    __in LONG64 lsn,
    __in PhysicalLogRecord & invalidPhysicalLogRecord,
    __in TransactionLogRecord & invalidTransactionLog,
    __in KAllocator & allocator)
{
    OperationLogRecord* pointer = _new(OPERATIONLOGRECORD_TAG, allocator) OperationLogRecord(
        recordType, 
        recordPosition, 
        lsn, 
        invalidPhysicalLogRecord, 
        invalidTransactionLog);

    THROW_ON_ALLOCATION_FAILURE(pointer);
    return OperationLogRecord::SPtr(pointer);
}

OperationLogRecord::SPtr OperationLogRecord::Create(
    __in TransactionBase & transaction,
    __in_opt OperationData const * const metaData,
    __in_opt OperationData const * const undo,
    __in_opt OperationData const * const redo,
    __in_opt OperationContext const * const operationContext,
    __in PhysicalLogRecord & invalidPhysicalLogRecord,
    __in TransactionLogRecord & invalidTransactionLog,
    __in KAllocator & allocator)
{
    OperationLogRecord* pointer = _new(OPERATIONLOGRECORD_TAG, allocator) OperationLogRecord(
        transaction,
        metaData,
        undo,
        redo,
        operationContext,
        invalidPhysicalLogRecord,
        invalidTransactionLog);

    THROW_ON_ALLOCATION_FAILURE(pointer);
    return OperationLogRecord::SPtr(pointer);
}

OperationLogRecord::SPtr OperationLogRecord::Create(
    __in TransactionBase & transaction,
    __in_opt OperationData const * const metaData,
    __in_opt OperationData const * const redo,
    __in_opt OperationContext const * const operationContext,
    __in PhysicalLogRecord & invalidPhysicalLogRecord,
    __in TransactionLogRecord & invalidTransactionLog,
    __in KAllocator & allocator)
{
    OperationLogRecord* pointer = _new(OPERATIONLOGRECORD_TAG, allocator) OperationLogRecord(
        transaction,
        metaData,
        redo,
        operationContext,
        invalidPhysicalLogRecord,
        invalidTransactionLog);

    THROW_ON_ALLOCATION_FAILURE(pointer);
    return OperationLogRecord::SPtr(pointer);
}

std::wstring OperationLogRecord::ToString() const
{
    wstring boolean = isRedoOnly_ ? L"true" : L"false"; 

    std::wstring logRecordString = Constants::CompEndlJSON;

    logRecordString += L"Redo Only" + Constants::DivisionBoolJSON + boolean;
    logRecordString += Constants::CloseJSON;;

    return __super::ToString() + logRecordString;
}


void OperationLogRecord::UpdateApproximateDiskSize()
{
    ApproximateSizeOnDisk = ApproximateSizeOnDisk + DiskSpaceUsed;
    ApproximateSizeOnDisk = ApproximateSizeOnDisk + __super::CalculateDiskWriteSize(metaData_.RawPtr());
    ApproximateSizeOnDisk = ApproximateSizeOnDisk + __super::CalculateDiskWriteSize(redo_.RawPtr());

    if (!isRedoOnly_)
    {
        ApproximateSizeOnDisk = ApproximateSizeOnDisk + __super::CalculateDiskWriteSize(undo_.RawPtr());
    }
}

OperationContext::CSPtr OperationLogRecord::ResetOperationContext()
{
    OperationContext::CSPtr context = operationContext_;
    if (context != nullptr)
    {
        operationContext_ = nullptr;
    }

    return context;
}

void OperationLogRecord::Read(
    __in BinaryReader & binaryReader,
    __in bool isPhysicalRead)
{
    // IF THIS METHOD CHANGES, READLOGICAL MUST CHANGE ACCORDINGLY AS WE ARE ADDING MULTIPLE ARRAYSEGMENTS HERE
    // THIS IS TRUE FOR ANY REDO+UNDO TYPE RECORDS
    __super::Read(binaryReader, isPhysicalRead);

    // Read Metadata size
    ULONG32 logicalStartingPosition = binaryReader.Position;
    ULONG32 logicalSizeOfSection = 0;
    binaryReader.Read(logicalSizeOfSection);
    ULONG32 logicalEndPosition = logicalStartingPosition + logicalSizeOfSection;

    binaryReader.Read(isRedoOnly_);

    // Jump to the end of the section ignoring fields that are not understood.
    ASSERT_IFNOT(
        logicalEndPosition >= binaryReader.Position,
        "Invalid operation log record, end position: {0} reader position:{1}",
        logicalEndPosition, binaryReader.Position);

    binaryReader.Position = logicalEndPosition;

    if (isRedoOnly_)
    {
        transactionBase_ = AtomicRedoOperation::CreateAtomicRedoOperation(
            transactionBase_->TransactionId,
            false,
            GetThisAllocator()).RawPtr();
    }

    metaData_ = OperationData::DeSerialize(binaryReader, GetThisAllocator());
    redo_ = OperationData::DeSerialize(binaryReader, GetThisAllocator());

    if (!isRedoOnly_)
    {
        undo_ = OperationData::DeSerialize(binaryReader, GetThisAllocator());
    }

    UpdateApproximateDiskSize();
}

void OperationLogRecord::ReadLogical(
    __in OperationData const & operationData,
    __inout INT & index)
{
    __super::ReadLogical(operationData, index);
    
    KBuffer::CSPtr buffer = OperationData::IncrementIndexAndGetBuffer(operationData, index); 
    BinaryReader reader(*buffer, GetThisAllocator());

    // Read Metadata size
    ULONG32 logicalStartingPosition = reader.Position;
    ULONG32 logicalSizeOfSection = 0;
    reader.Read(logicalSizeOfSection);
    ULONG32 logicalEndPosition = logicalStartingPosition + logicalSizeOfSection;

    reader.Read(isRedoOnly_);

    // Jump to the end of the section ignoring fields that are not understood.
    ASSERT_IFNOT(
        logicalEndPosition >= reader.Position,
        "Invalid operation log record, logical end position: {0} reader position:{1}",
        logicalEndPosition, reader.Position);

    reader.Position = logicalEndPosition;

    if (isRedoOnly_)
    {
        transactionBase_ = AtomicRedoOperation::CreateAtomicRedoOperation(
            transactionBase_->TransactionId,
            false,
            GetThisAllocator()).RawPtr();
    }

    metaData_ = OperationData::DeSerialize(operationData, index, GetThisAllocator());
    redo_ = OperationData::DeSerialize(operationData, index, GetThisAllocator());
   
    if (!isRedoOnly_)
    {
        undo_ = OperationData::DeSerialize(operationData, index, GetThisAllocator());
    }

    UpdateApproximateDiskSize();
}

void OperationLogRecord::Write(
    __in BinaryWriter & binaryWriter,
    __inout OperationData & operationData,
    __in bool isPhysicalWrite,
    __in bool forceRecomputeOffsets)
{
    __super::Write(binaryWriter, operationData, isPhysicalWrite, forceRecomputeOffsets);

    if (replicatedData_ == nullptr)
    {
        OperationData::SPtr localReplicatedData = OperationData::Create(GetThisAllocator());
        ULONG32 startingPosition = binaryWriter.Position;

        // Metadata Size
        binaryWriter.Position += sizeof(ULONG32);

        binaryWriter.Write(isRedoOnly_);

        // End of metadata.
        ULONG32 endPosition = binaryWriter.Position;
        ULONG32 sizeOfSection = endPosition - startingPosition;

        binaryWriter.Position = startingPosition;
        binaryWriter.Write(sizeOfSection);
        binaryWriter.Position = endPosition;

        KBuffer::SPtr logMetaData = binaryWriter.GetBuffer(startingPosition);
        localReplicatedData->Append(*logMetaData);

        // Write user metadata
        OperationData::Serialize(
            binaryWriter,
            metaData_.RawPtr(),
            *localReplicatedData);

        // Write user redo data
        OperationData::Serialize(
            binaryWriter,
            redo_.RawPtr(),
            *localReplicatedData);

        if (!isRedoOnly_)
        {
            // Write user undo data
            OperationData::Serialize(
                binaryWriter,
                undo_.RawPtr(),
                *localReplicatedData);
        }

        replicatedData_ = localReplicatedData.RawPtr();
    }

    operationData.Append(*replicatedData_);
}

ULONG OperationLogRecord::GetSizeOnWire() const
{
    ULONG size = 
        __super::GetSizeOnWire() +
        sizeof(LONG32) +
        __super::CalculateWireSize(metaData_.RawPtr()) +
        __super::CalculateWireSize(redo_.RawPtr()) +
        sizeof(bool);

    if (!isRedoOnly_)
    {
        size += __super::CalculateWireSize(undo_.RawPtr());
    }

    return size;
}

bool OperationLogRecord::Test_Equals(__in LogRecord const & other) const
{
    OperationLogRecord const & otherOperationLogRecord = dynamic_cast<OperationLogRecord const &>(other);

    if (__super::Test_Equals(other))
    {
        return
            isRedoOnly_ == otherOperationLogRecord.isRedoOnly_ &&
            metaData_ != nullptr ?
                metaData_->Test_Equals(*otherOperationLogRecord.metaData_) :
                (metaData_ == otherOperationLogRecord.metaData_) != 0 &&
            redo_ != nullptr ?
                redo_->Test_Equals(*otherOperationLogRecord.redo_) :
                (redo_ == otherOperationLogRecord.redo_) != 0 &&
            undo_ != nullptr ?
                undo_->Test_Equals(*otherOperationLogRecord.undo_) :
                (undo_ == otherOperationLogRecord.undo_) != 0;
    }

    return false;
}
